AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/AWSIncomingEventReconciliationQueue.swift

```diff
@@ -87,6 +91,30 @@ final class AWSIncomingEventReconciliationQueue: IncomingEventReconciliationQueu
             self.eventReconciliationQueueTopic.send(.mutationEvent(event))
         }
     }
+
+    private func kill() -> Bool {
+
+        //It is possible to have multiple persistent connections fail in a very
+        // narrow time window, which will result in having multiple calls to kill.
+        // We should avoid calling kill() multiple times on items that have already
+        // been killed.
+        let semaphore = DispatchSemaphore(value: 1)
+        semaphore.wait()
+        let shouldIssueKill = !hasIssuedKillSequence
+        hasIssuedKillSequence = true
+        semaphore.signal()
+
+        if shouldIssueKill {
+            modelReconciliationQueueSinks.values.forEach { $0.cancel() }
+            reconciliationQueues.values.forEach { $0.kill() }
+            reconciliationQueues = [:]
+            modelReconciliationQueueSinks = [:]
+            return true
+        } else {
+            return false
+        }
```

Comment:
Not super proud of this.  I would have liked to receive one and only one message from any downstream recieveCompletion() on any single model reconciliation queue.  Couldn't figure out a simpler way to do this, so just used a simple semaphore to do this only once.

Reply:
I may be misunderstanding the problem, but I think what you want is to [`merge()`](https://developer.apple.com/documentation/combine/publisher/3204724-merge) the downstream publishers, and wait for completions in this class' subscriber. The `kill()` method should simply iterate over each queue as you're doing and kill it, but don't cancel the sink. When each reconciliationQueue emits a completion, the merged sink will receive its own completion, which you can then use to trigger the remaining cleanup, including notifying upstreams by emitting your own completion. Sample playground code: https://gist.github.com/palpatim/057856de01a2c0202ef6b7b4c02bff8d

Reply:
So the problem this addresses is that one or more reconciliation queues have the ability to send a completion on their publisher, which is received by a subscriber -- in this case, this class: `AWSIncomingEventReconciliationQueue`.  For example, we have 27 persistent websockets (because we have 9 syncable models), and 5 of those websockets fail -- this would mean `AWSIncomingEventReconciliationQueue:: onReceiveCompletion()` would get called 5 times -- meaning `AWSIncomingEventReconciliationQueue:: kill()->Bool` would get called 5 times, which means we'd attempt to kill reconciliation queues that potentially are either already killed, or in the process of being killed.  This has lead to crashes, and (somehow?) memory leaks.  Therefore, I've created this Semaphore which locks out access to other executing threads.